scripts/remotes.lua

```
@@ -1,13 +1,217 @@
 -- Copyright (c) 2020 Dockmeister
 -- Copyright (c) 2023 Branko Majic
+-- Copyright (c) 2024 Bryan O'Malley
 -- Provided under MIT license. See LICENSE for details.
 
 
 -- Main implementation
 -- ===================
 
 local remotes = {}
+local overlap_tolerance = 0.9
 
+--- Calculates square distance between 2 positions
+--
+-- @param a first position
+-- @param b second position--
+--
+-- @return squared distance between the positions
+--
+function dist_sq(a, b)
+  distance = (a.x - b.x) ^ 2 + (a.y - b.y) ^ 2
+  return distance
+end
+
+--- Calculates the center of the passed in circle or rectangle
+--
+-- @param shape the shape
+--
+-- @return the center of the shape
+--
+function shape_center(shape)
+  if shape.center then
+    return shape.center
+  else
+    return {x = 0.5 * (shape.left_top.x + shape.right_bottom.x), y = 0.5 * (shape.left_top.y + shape.right_bottom.y)}
+  end
+end
+
+--- Calculates the approximate center point between an array of circles
+--
+-- Calculation is a weighted average of the circle centers based on inverse radii weights (closer to smaller circles)
+--
+-- @param targets the array of circles
+--
+-- @return the approximate center point between the circles
+--
+function circle_centroid(targets)
+  local average = { x = 0.0, y = 0.0 }
+  local length = #targets
+  if length == 0 then
+    return average
+  end
+  -- approximate finding center point between circle edges
+  local weight = 0
+  for _, target in pairs(targets) do
+    if target.radius == nil then
+      return average -- quit early if we've not been passed all circles
```

Should this ever happen? Would the resulting average have any meaningful use or would it actually be invalid? If the code should never end up in this situation, I would probably drop this check and let it crash on accessing `.radius` below (since it might hide issues elsewhere in the code).

---

Assuming you'd already processed some of the shapes it'd at least have that value, but may as well let it fail.